### PR TITLE
PPM-334 slave: handle_1905_higher_layer_data_message: move from backhaul to son_slave

### PR DESCRIPTION
**Pull request by orenvor**
Move handling of 1905_higher_layer_data messages from the backhual manager to the son_slave thread.

https://jira.prplfoundation.org/browse/PPM-334

### CODE REVIEW

File: agent/src/beerocks/slave/son_slave_thread.cpp [hunk 1]
```diff
@@ -40,6 +40,7 @@
 #include <tlvf/wfa_map/tlvChannelSelectionResponse.h>
 #include <tlvf/wfa_map/tlvClientAssociationControlRequest.h>
 #include <tlvf/wfa_map/tlvClientAssociationEvent.h>
+#include <tlvf/wfa_map/tlvHigherLayerData.h>
```
Comment (adam1985d) [inline agent/src/beerocks/slave/son_slave_thread.cpp:43]: Please also remove this include from the `backhaul_manager.cpp` file if possible

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 2]
```diff
@@ -2321,9 +2321,6 @@ bool backhaul_manager::handle_1905_1_message(ieee1905_1::CmduMessageRx &cmdu_rx,
     case ieee1905_1::eMessageType::AP_CAPABILITY_QUERY_MESSAGE: {
         return handle_ap_capability_query(cmdu_rx, src_mac);
     }
```
Comment (adam1985d) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:2092]: the commit message should follow the `contribution.md` rules.

Please use this PR to ramp-up on them.
(just from a quick review: line-length limits of title and body are not met)

File: agent/src/beerocks/slave/son_slave_thread.cpp [hunk 3]
```diff
@@ -4216,6 +4219,38 @@ bool slave_thread::handle_client_association_request(Socket *sd, ieee1905_1::Cmd
     return send_cmdu_to_controller(cmdu_tx);
 }
 
+bool slave_thread::handle_1905_higher_layer_data_message(Socket *sd,
+                                                         ieee1905_1::CmduMessageRx &cmdu_rx)
+{
+    //only one backhaul manager (the slave) should return ACK for higher layer data message.
```
Comment (adam1985d) [inline agent/src/beerocks/slave/son_slave_thread.cpp]:
```suggestion
    // Only one backhaul manager (the slave) should return ACK for higher layer data message.
```
or anything else you think can make it clearer.

In general - although it is porting existing code from point A to point B, you can always rework is where you find it suitable.
You can use there small PRs to ramp on the coding guidelines and enforce them to the ported code.

File: agent/src/beerocks/slave/son_slave_thread.cpp [hunk 4]
```diff
@@ -4216,6 +4219,38 @@ bool slave_thread::handle_client_association_request(Socket *sd, ieee1905_1::Cmd
...
+    auto tlvHigherLayerData = cmdu_rx.getClass<wfa_map::tlvHigherLayerData>();
+    if (!tlvHigherLayerData) {
+        LOG(ERROR) << "addClass wfa_map::tlvHigherLayerData failed";
+        return false;
+    }
+
+    const auto protocol       = tlvHigherLayerData->protocol();
+    const auto payload_length = tlvHigherLayerData->payload_length();
+    LOG(DEBUG) << "protocol: " << std::hex << int(protocol);
+    LOG(DEBUG) << "payload_length: " << std::hex << int(payload_length);
```
Comment (adam1985d) [inline agent/src/beerocks/slave/son_slave_thread.cpp]:
```suggestion
    LOG(DEBUG) << "Protocol: " << std::hex << int(protocol);
    LOG(DEBUG) << "Payload-Length: " << std::hex << int(payload_length);
```
we should print the meaning of the data printed instead of the parameter name. this also makes future naming changes (that do not change the meaning) not required log print changes.

that being said:
1. please verify if the "payload_length" print is expected by some certification test before you change the print to "Payload-Length" and "Protocol"
2. if you change this and the certification tests don't use this print as reference (only our test-flows) then you must align our test-flows to the new print

File: agent/src/beerocks/slave/son_slave_thread.cpp [hunk 5]
```diff
@@ -4216,6 +4219,38 @@ bool slave_thread::handle_client_association_request(Socket *sd, ieee1905_1::Cmd
...
+
+    // build ACK message CMDU
```
Comment (adam1985d) [inline agent/src/beerocks/slave/son_slave_thread.cpp]:
```suggestion
    // Build ACK message CMDU
```

File: agent/src/beerocks/slave/son_slave_thread.cpp [hunk 6]
```diff
@@ -4216,6 +4219,38 @@ bool slave_thread::handle_client_association_request(Socket *sd, ieee1905_1::Cmd
...
+    auto cmdu_tx_header = cmdu_tx.create(mid, ieee1905_1::eMessageType::ACK_MESSAGE);
+    if (!cmdu_tx_header) {
+        LOG(ERROR) << "cmdu creation of type ACK_MESSAGE, has failed";
+        return false;
+    }
+    LOG(DEBUG) << "sending ACK message to the originator, mid=" << std::hex << int(mid);
```
Comment (adam1985d) [inline agent/src/beerocks/slave/son_slave_thread.cpp]:
```suggestion
    LOG(DEBUG) << "Sending ACK message to the originator, mid=" << std::hex << int(mid);
```

File: tests/test_flows.py [hunk 7]
```diff
@@ -44,14 +44,17 @@ def start_test(self, test: str):
         status(test + " starting")
 
     def check_log(self, entity_or_radio: Union[env.ALEntity, env.Radio], regex: str,
-                  start_line: int = 0) -> bool:
-        '''Verify that the logfile for "entity_or_radio" matches "regex", fail if not.'''
-        return self.wait_for_log(entity_or_radio, regex, start_line, 0.3)
+                  start_line: int = 0, fail_when_not_found: bool = True) -> bool:
+        '''Verify that the logfile for "entity_or_radio" matches "regex",
+           fail if no match when fail_when_not_found is enabled.
+        '''
```
Comment (adam1985d) [inline tests/test_flows.py]:
```suggestion
        '''Verify that the log-file for "entity_or_radio" matches "regex",
           fail if no match is found when fail_when_not_found is enabled.
        '''
```

File: tests/test_flows.py [hunk 8]
```diff
@@ -1246,14 +1249,25 @@ def test_higher_layer_data_payload_trigger(self):
...
+        received_in_radio0, _, _ = self.check_log(env.agents[0].radios[0],
+                                                  r"HIGHER_LAYER_DATA_MESSAGE",
+                                                  fail_when_not_found=False)
+        received_in_radio1, _, _ = self.check_log(env.agents[0].radios[1],
+                                                  r"HIGHER_LAYER_DATA_MESSAGE",
+                                                  fail_when_not_found=False)
+        if (received_in_radio0 == received_in_radio1):
+            self.fail(f"higher layer data message was received and acknowledged by "
+                      f"{int(received_in_radio0) + int(received_in_radio1)} agent's radios, "
+                      f"expected only 1")
```
Comment (adam1985d) [inline tests/test_flows.py]:
```suggestion
                      f"expected exactly 1")
```
1. Since you decided to include in the same condition:
false,false
and
true,true
then your comment should reflect that.

2. This is a test so I would consider the following comment a nitpick - but:
I prefer the condition and the error-condition to be pre-acquired to the point you use them.
Meaning:
- Calculate the number of agents that received the message
- Then, compare it against the expected value.
- Then, print the expected and actual as part of the error log print

File: tests/test_flows.py [hunk 9]
```diff
@@ -1246,14 +1249,25 @@ def test_higher_layer_data_payload_trigger(self):
...
 
-        debug("Confirming matching protocol and payload length")
+        received_agent_radio = (env.agents[0].radios[0] if received_in_radio0
+                                else env.agents[0].radios[1])
 
-        self.check_log(env.agents[0], r"protocol: 0")
-        self.check_log(env.agents[0], r"payload_length: 0x4b0")
+        debug("Confirming matching protocol and payload length")
+        self.check_log(received_agent_radio, r"protocol: 0")
+        self.check_log(received_agent_radio, r"payload_length: 0x4b0")
```
Comment (adam1985d) [inline tests/test_flows.py]: please verify if the "payload_length" print is expected by some certification test before you change the print to "Payload-Length" and "Protocol"

File: tests/test_flows.py [hunk 10]
```diff
@@ -1246,14 +1249,25 @@ def test_higher_layer_data_payload_trigger(self):
...
+                                                  r"HIGHER_LAYER_DATA_MESSAGE",
+                                                  fail_when_not_found=False)
+        if (received_in_radio0 == received_in_radio1):
+            self.fail(f"higher layer data message was received and acknowledged by "
+                      f"{int(received_in_radio0) + int(received_in_radio1)} agent's radios, "
+                      f"expected only 1")
 
-        debug("Confirming matching protocol and payload length")
+        received_agent_radio = (env.agents[0].radios[0] if received_in_radio0
+                                else env.agents[0].radios[1])
```
Comment (adam1985d) [inline tests/test_flows.py:1268]: do we care for debug purposes on which radio we received the message?
if yes, then need to add debug print regarding this.
@morantr

**Review by adam1985d — CHANGES_REQUESTED** [review]
In general, I would have approved this PR (since the changes are small and don't require a re-review).
That being said - since these are your first PRs, I want to re-review to make sure the review-and-delivery-process is met.

File: tests/test_flows.py [hunk 11]
```diff
@@ -44,14 +44,17 @@ def start_test(self, test: str):
         status(test + " starting")
 
     def check_log(self, entity_or_radio: Union[env.ALEntity, env.Radio], regex: str,
-                  start_line: int = 0) -> bool:
-        '''Verify that the logfile for "entity_or_radio" matches "regex", fail if not.'''
-        return self.wait_for_log(entity_or_radio, regex, start_line, 0.3)
+                  start_line: int = 0, fail_when_not_found: bool = True) -> bool:
+        '''Verify that the logfile for "entity_or_radio" matches "regex",
+           fail if no match when fail_when_not_found is enabled.
```
Comment (morantr) [inline tests/test_flows.py]: I would change `fail_when_not_found ` to `fail_on_mistmatch".
Also, please wrap the argument name on the comments with double-quotes to make it clear that it is an argument and not regular text.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 2 above)
Comment (morantr) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:2092]: Also, the commit message is not accurate.
It is not that the handler should be on son_slave. It should be on the future Unified Agent, which the son_slave will be converted to.

I noticed a typo: "...match the new **behiver**" -> "...match the new **behavior**".
I really recommend you, to install `Grammarly` extension to chrome, and past there the commit message before committing. We are not native English speakers, and this tool can save you a lot of rework.

Last thing. You don't need to supply full Jira link, `PPM-334` is enough.

File: tests/test_flows.py (hunk 10 above)
Comment (morantr) [reply tests/test_flows.py:1268]: It is used for the next check on the log files. Once we know which son_slave received it, we don't need to check on both of them.

File: tests/test_flows.py (hunk 9 above)
Comment (morantr) [reply tests/test_flows.py]: Certification tests don't scan log files, only our internal `test_flows` script.

**Review by morantr — APPROVED** [review]
Good PR.
I would split the change on `test_flows` to a second commit and explaining on the commit message that due to the change on the handler location, the `test_flow` will break and will be fixed on the next commit.
But since you already did it like this, it is fine by me to leave it that way.

File: agent/src/beerocks/slave/son_slave_thread.cpp (hunk 1 above)
Comment (orenvor) [reply agent/src/beerocks/slave/son_slave_thread.cpp:43]: include removed from `backhaul_manager.cpp`.

File: agent/src/beerocks/slave/son_slave_thread.cpp (hunk 5 above)
Comment (orenvor) [reply agent/src/beerocks/slave/son_slave_thread.cpp]: fixed

File: agent/src/beerocks/slave/son_slave_thread.cpp (hunk 6 above)
Comment (orenvor) [reply agent/src/beerocks/slave/son_slave_thread.cpp]: fixed

File: tests/test_flows.py (hunk 7 above)
Comment (orenvor) [reply tests/test_flows.py]: fixed

File: tests/test_flows.py (hunk 8 above)
Comment (orenvor) [reply tests/test_flows.py]: suggestion applied

File: tests/test_flows.py (hunk 9 above)
Comment (orenvor) [reply tests/test_flows.py]: suggestion applied.

File: tests/test_flows.py (hunk 11 above)
Comment (orenvor) [reply tests/test_flows.py]: applied suggestion.

File: agent/src/beerocks/slave/son_slave_thread.cpp (hunk 4 above)
Comment (orenvor) [reply agent/src/beerocks/slave/son_slave_thread.cpp]: suggestion applied.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 2 above)
Comment (orenvor) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:2092]: fixed

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 12]
```diff
@@ -83,7 +83,6 @@
 #include <tlvf/wfa_map/tlvClientCapabilityReport.h>
 #include <tlvf/wfa_map/tlvClientInfo.h>
 #include <tlvf/wfa_map/tlvErrorCode.h>
-#include <tlvf/wfa_map/tlvHigherLayerData.h>
 #include <tlvf/wfa_map/tlvMetricReportingPolicy.h>
```
Comment (adam1985d) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:86]: the commit title is still too long (see contribution.md):
you can replace with:
```
agent: move 1905_higher_layer_data handling to son_slave
```

**Review by adam1985d — APPROVED** [review]
I approved but please change the commit title to meet the length limit

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 12 above)
Comment (morantr) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:86]: The commits message is ok. As long there is no `...` at the end, it is ok.

File: agent/src/beerocks/slave/son_slave_thread.cpp [hunk 13]
```diff
@@ -4198,6 +4201,38 @@ bool slave_thread::handle_client_association_request(Socket *sd, ieee1905_1::Cmd
     return send_cmdu_to_controller(cmdu_tx);
 }
 
+bool slave_thread::handle_1905_higher_layer_data_message(Socket *sd,
+                                                         ieee1905_1::CmduMessageRx &cmdu_rx)
+{
+    // Only one backhaul manager (the slave) should return ACK for higher layer data message.
+    if (is_backhaul_manager) {
```
Comment (VladyslavTupikin) [inline agent/src/beerocks/slave/son_slave_thread.cpp:4208]: Do we need this check here?

File: tests/test_flows.py [hunk 14]
```diff
@@ -44,14 +44,17 @@ def start_test(self, test: str):
         status(test + " starting")
 
     def check_log(self, entity_or_radio: Union[env.ALEntity, env.Radio], regex: str,
-                  start_line: int = 0) -> bool:
-        '''Verify that the logfile for "entity_or_radio" matches "regex", fail if not.'''
-        return self.wait_for_log(entity_or_radio, regex, start_line, 0.3)
+                  start_line: int = 0, fail_on_mistmatch: bool = True) -> bool:
```
Comment (RanRegev) [inline tests/test_flows.py]: I don't understand this change:
changes to `check_log` should be done in a separate commit, with its own rational - as it serves other tests as well.

File: agent/src/beerocks/slave/son_slave_thread.h [hunk 15]
```diff
@@ -234,6 +234,7 @@ class slave_thread : public beerocks::socket_thread {
     beerocks::message::sWifiChannel channel_selection_select_channel();
     bool handle_multi_ap_policy_config_request(Socket *sd, ieee1905_1::CmduMessageRx &cmdu_rx);
     bool handle_client_association_request(Socket *sd, ieee1905_1::CmduMessageRx &cmdu_rx);
+    bool handle_1905_higher_layer_data_message(Socket *sd, ieee1905_1::CmduMessageRx &cmdu_rx);
```
Comment (VladyslavTupikin) [inline agent/src/beerocks/slave/son_slave_thread.h]:
```suggestion
    bool handle_1905_higher_layer_data_message(Socket &sd, ieee1905_1::CmduMessageRx &cmdu_rx);
```

File: agent/src/beerocks/slave/son_slave_thread.cpp [hunk 16]
```diff
@@ -4198,6 +4201,38 @@ bool slave_thread::handle_client_association_request(Socket *sd, ieee1905_1::Cmd
     return send_cmdu_to_controller(cmdu_tx);
 }
 
+bool slave_thread::handle_1905_higher_layer_data_message(Socket *sd,
```
Comment (VladyslavTupikin) [inline agent/src/beerocks/slave/son_slave_thread.cpp]:
```suggestion
bool slave_thread::handle_1905_higher_layer_data_message(Socket &sd,
```

File: tests/test_flows.py (hunk 14 above)
Comment (VladyslavTupikin) [inline tests/test_flows.py]: Changes in test_flows should be done in another commit.

File: tests/test_flows.py [hunk 17]
```diff
@@ -1246,14 +1249,27 @@ def test_higher_layer_data_payload_trigger(self):
         mid = env.controller.dev_send_1905(env.agents[0].mac, 0x8018,
                                            tlv(0xA0, 0x04b1, "{0x00 %s}" % payload))
 
-        debug("Confirming higher layer data message was received in the agent")
+        debug("Confirming higher layer data message was received in one of the agent's radios")
 
-        self.check_log(env.agents[0], r"HIGHER_LAYER_DATA_MESSAGE")
+        received_in_radio0, _, _ = self.check_log(env.agents[0].radios[0],
```
Comment (RanRegev) [inline tests/test_flows.py:1254]: A comment only:
We should stop using check_log to verify correctness.
We can either:
* query the application (best)
* check packets (good)
I leave it as a comment only because this is not a new `check_log`.

File: controller/src/beerocks/master/son_master_thread.cpp [hunk 18]
```diff
@@ -1509,8 +1509,8 @@ bool master_thread::handle_cmdu_1905_higher_layer_data_message(const std::string
 
     const auto protocol       = tlvHigherLayerData->protocol();
     const auto payload_length = tlvHigherLayerData->payload_length();
-    LOG(DEBUG) << "protocol: " << std::hex << protocol;
-    LOG(DEBUG) << "payload length: " << int(payload_length);
+    LOG(DEBUG) << "Protocol: " << std::hex << protocol;
```
Comment (VladyslavTupikin) [inline controller/src/beerocks/master/son_master_thread.cpp:1512]: These changes should be done in another commit.

**Review by VladyslavTupikin — CHANGES_REQUESTED** [review]

**Review by RanRegev — APPROVED** [review]
Approving, 
though  I would like to see the changes to `check_log` function introduced in a different PR.

File: agent/src/beerocks/slave/son_slave_thread.cpp (hunk 13 above)
Comment (morantr) [reply agent/src/beerocks/slave/son_slave_thread.cpp:4208]: Yes.

File: agent/src/beerocks/slave/son_slave_thread.h (hunk 15 above)
Comment (orenvor) [reply agent/src/beerocks/slave/son_slave_thread.h]: suggestion applied.

File: agent/src/beerocks/slave/son_slave_thread.cpp (hunk 16 above)
Comment (orenvor) [reply agent/src/beerocks/slave/son_slave_thread.cpp]: suggestion applied.

File: tests/test_flows.py (hunk 14 above)
Comment (orenvor) [reply tests/test_flows.py]: fixed.

File: controller/src/beerocks/master/son_master_thread.cpp (hunk 18 above)
Comment (orenvor) [reply controller/src/beerocks/master/son_master_thread.cpp:1512]: fixed.

File: tests/test_flows.py (hunk 14 above)
Comment (orenvor) [reply tests/test_flows.py]: Currently, the `check_log`  function is used as an "assert" statement, 
designed to fail the test when a string was not found in the log.

This change allows testing of the logs without the "assert" part, 
for cases where a missing string is expected or acceptable.

**Review by VladyslavTupikin — APPROVED** [review]
